message/deduplication/ephemeralbloomfilter/filter.go

```diff
@@ -0,0 +1,80 @@
+package ephemeralbloomfilter
+
+import (
+	"github.com/bits-and-blooms/bloom/v3"
+)
+
+type options struct {
+	estimatedItems    uint
+	falsePositiveRate float64
+}
+
+type optionsFunc struct {
+	f func(o *options)
+}
+
+func (f optionsFunc) apply(o *options) {
+	f.f(o)
+}
+
+type Option interface {
+	apply(*options)
+}
+
+// WithEstimatedItems sets the estimated items expected for the lifetime of the
+// bloom filter.
+func WithEstimatedItems(items uint) Option {
+	return optionsFunc{
+		f: func(o *options) {
+			o.estimatedItems = items
+		},
+	}
+}
+
+// WithFalsePositiveRate sets the desired false positive rate in the bloom
+// filter which lets the implementation decide how many hashing algorithms and
+// bit array size to use.
+func WithFalsePositiveRate(rate float64) Option {
+	return optionsFunc{
+		f: func(o *options) {
+			o.falsePositiveRate = rate
+		},
+	}
+}
+
+type backend struct {
+	bf *bloom.BloomFilter
+}
+
+func (e *backend) Set(key string) error {
+	e.bf.AddString(key)
+
+	return nil
+}
+
+func (e *backend) Exists(key string) (bool, error) {
+	return e.bf.TestString(key), nil
+}
+
+func (e *backend) CheckAndSet(key string) (bool, error) {
+	return e.bf.TestAndAddString(key), nil
+}
+
+// New returns a implementation if
+// bloomFilterBackend that will not be persisted into storage.
+func New(opts ...Option) *backend {
```

Comment:
Is there a reason to return a non-exported type here? It seems like it would make it hard to use because you can only use an interface type to store it, not the struct type itself? We also don't export the interface type it seems like...

Reply:
There's no reason to export it. The constructor is exported and that's all you need because of the options variadic. This backend is tailored for `deduplication.BloomFilter`'s use and if someone really wants to use the data structure inside this type, they should just use github.com/bits-and-blooms/bloom directly. I think idealy these backends should be in an `internal` directory, but that complicates bootstrapping these with wire and stuff, so I just make everything unexported except the Option interface and constructor to discourage out of scope use.